src/core/server/saved_objects/service/lib/update_objects_spaces.ts

```diff
@@ -0,0 +1,310 @@
+        spacesToRemove
+      );
+      const expectedResult = {
+        type,
+        id,
+        remainingSpaces,
+        ...(isUpdateRequired && { esRequestIndex: bulkOperationRequestIndexCounter++ }),
+      };
+
+      if (isUpdateRequired) {
+        const documentMetadata = {
+          _id: serializer.generateRawId(undefined, type, id),
+          _index: getIndexForType(type),
+          ...versionProperties,
+        };
+        if (remainingSpaces.length) {
+          const documentToSave = { updated_at: time, namespaces: remainingSpaces };
+          // @ts-expect-error BulkOperation.retry_on_conflict, BulkOperation.routing. BulkOperation.version, and BulkOperation.version_type are optional
+          bulkOperationParams.push({ update: documentMetadata }, { doc: documentToSave });
+        } else {
+          // @ts-expect-error BulkOperation.retry_on_conflict, BulkOperation.routing. BulkOperation.version, and BulkOperation.version_type are optional
+          bulkOperationParams.push({ delete: documentMetadata });
+        }
+      }
+
+      return { tag: 'Right' as 'Right', value: expectedResult };
+    }
+  );
+
+  const { refresh = DEFAULT_REFRESH_SETTING } = options;
+  const bulkOperationResponse = bulkOperationParams.length
+    ? await client.bulk({ refresh, body: bulkOperationParams, require_alias: true })
+    : undefined;
+
+  return {
+    objects: expectedBulkOperationResults.map<SavedObjectsUpdateObjectsSpacesResponseObject>(
+      (expectedResult) => {
+        if (isLeft(expectedResult)) {
+          return expectedResult.error;
+        }
+
+        const { type, id, remainingSpaces, esRequestIndex } = expectedResult.value;
+        if (esRequestIndex !== undefined) {
+          const response = bulkOperationResponse?.body.items[esRequestIndex] ?? {};
+          const rawResponse = Object.values(response)[0] as any;
+          const error = getBulkOperationError(type, id, rawResponse);
+          if (error) {
+            return { id, type, spaces: [], error };
+          }
+        }
+
+        return { id, type, spaces: remainingSpaces };
+      }
+    ),
+  };
+}
+
+/** Extracts the contents of a decorated error to return the attributes for bulk operations. */
+function errorContent(error: DecoratedError) {
+  return error.output.payload;
+}
+
+/** Gets the remaining spaces for an object after adding new ones and removing old ones. */
+function getRemainingSpaces(
+  existingSpaces: string[],
+  spacesToAdd: string[],
+  spacesToRemove: string[]
+) {
+  const addSet = new Set(spacesToAdd);
+  const removeSet = new Set(spacesToRemove);
+  const remainingSpaces = existingSpaces
+    .filter((x) => {
+      addSet.delete(x);
+      return !removeSet.delete(x);
+    })
+    .concat(Array.from(addSet));
+
+  const isAnySpaceAdded = addSet.size > 0;
+  const isAnySpaceRemoved = removeSet.size < spacesToRemove.length;
+  const isUpdateRequired = remainingSpaces.length === 0 || isAnySpaceAdded || isAnySpaceRemoved;
```

Comment:
Is the `remainingSpaces.length === 0` condition necessary? Can we get to a state where `remainingSpaces.length === 0` while both `isAnySpaceAdded` and `isAnySpaceRemoved` are `false`?

Reply:
My reasoning for this was:

1. If an object exists in only "space-x", and you remove it from "space-x", it should be deleted
2. If an object doesn't exist in any spaces, and you remove it from "space-x", it should be deleted

This conditional is just some defensive coding for the second scenario. It could have happened in my original implementation of this method, which was not space-aware. However, I don't think that is a realistic scenario now, because this method _is_ space-aware (we use `rawDocExistsInNamespace` above); when we try to pre-fetch the object, if it doesn't exist in the current space it wouldn't ever reach this code path.

TL;DR I think we can remove this now for clarity, thanks for pointing it out! If we ever decide to make this method non-space-aware (so it will work the same no matter what your current space is), we will need to revisit this. Unfortunately, that is not possible right now because we don't have a way to check if the user is authorized to access an object type in "any space" (because of ES limitations, we can only check authorization for individual spaces or "all spaces").